### PR TITLE
method for reloading single node of DirectoryTree

### LINKED ISSUE #2757: Add a method to `DirectoryTree` that allows for reloading an individual node
The question cropped up on Discord and seems like a useful thing to have: the ability to say "clear out and reload this specific node" when using `DirectoryTree` -- a developer may know that a specific directory has been modified and they want to refresh the content.

### PR DESCRIPTION
**Pull request by gergely-elias**
**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

New method `DirectoryTree.reload_node` allows reloading the content of a single directory.

fixes #2757

### CODE REVIEW

File: tests/tree/test_directory_tree_reload_node.py
Comment (rodrigogiraoserrao) [inline tests/tree/test_directory_tree_reload_node.py:1]: This test looks great!

I'm just going to ask if we can avoid writing inside the files after creating them.
It saves us just a little bit of time! ⏩

File: tests/tree/test_directory_tree_reload_node.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+from rich.text import Text
+
+from textual.app import App, ComposeResult
+from textual.widgets import DirectoryTree
+
+
+class MyDirectoryTree(DirectoryTree):
```
Comment (rodrigogiraoserrao) [inline tests/tree/test_directory_tree_reload_node.py:7]: You can use `DirectoryTree` directly! 😃

**Review by rodrigogiraoserrao — CHANGES_REQUESTED** [review]
This is headed in the correct direction!
Just a couple of small changes, please!

**Comment by gergely-elias** [comment]
> This is headed in the correct direction! Just a couple of small changes, please!

Thanks for the comments, I'm about to push updates.
One thing I noticed is that there is a failing test case affecting tabs on windows/python3.8 setup.
I was wondering how stable the tests are and if I should look into that specific case.

File: src/textual/widgets/_directory_tree.py [hunk 2]
```diff
@@ -146,6 +146,56 @@ def reload(self) -> None:
         # loading up.
         self._add_to_load_queue(self.root)
 
+    def clear_subtree(self, node: TreeNode[DirEntry]) -> DirectoryTree:
```
Comment (davep) [inline src/textual/widgets/_directory_tree.py:149]: I think I'd have this typed to return `Self`.

File: src/textual/widgets/_directory_tree.py [hunk 3]
```diff
@@ -146,6 +146,56 @@ def reload(self) -> None:
...
+            node_data,
+            expanded=True,
+        )
+        self._updates += 1
+        self.refresh()
+        return self
+
+    def reset_subtree(
+        self, node: TreeNode[DirEntry], label: TextType, data: DirEntry | None = None
+    ) -> DirectoryTree:
```
Comment (davep) [inline src/textual/widgets/_directory_tree.py:173]: I think I'd have this typed to return `Self`.

**Comment by rodrigogiraoserrao** [comment]
@gergely-elias yeah some of our tests are a bit flaky; yesterday I noticed a failing test in this PR and it was completely unrelated to you. I re-ran it and it passed.
I'll let you take a look at @davep's comments and then I'll run CI again so we can merge your PR! 🚀

**Comment by gergely-elias** [comment]
Thanks @davep & @rodrigogiraoserrao.
I've pushed another commit, in which the import is behind an if - just like the way it is in the parent class.
But I saw that in other classes you import it without such condition, and it wasn't clear if there is a reason for this.

Let me know if it is fine as it is now or should be reworked.

**Review by rodrigogiraoserrao — APPROVED** [review]
Looks good to me! 🚀

**Review by willmcgugan — CHANGES_REQUESTED** [review]
It's probably best to add some utility methods on the base case as needed here.

File: src/textual/widgets/_directory_tree.py [hunk 4]
```diff
@@ -146,6 +149,56 @@ def reload(self) -> None:
...
+        """Clear all nodes under the given node.
+
+        Returns:
+            The `Tree` instance.
+        """
+        self._line_cache.clear()
+        self._tree_lines_cached = None
+        node_label = node._label
+        node_data = node.data
+        node = TreeNode(
```
[the remark this answers is not in the document]
Comment (gergely-elias) [reply src/textual/widgets/_directory_tree.py:162]: you're right, the second argument should be the parent node, and not `None`

File: tests/tree/test_directory_tree_reload_node.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    def __init__(self, path):
+        super().__init__()
+        self._tmp_path = path
+
+    def compose(self) -> ComposeResult:
+        yield DirectoryTree(self._tmp_path)
+
+
+async def test_directory_tree_reload_node(tmp_path) -> None:
```
[the remark this answers is not in the document]
Comment (gergely-elias) [reply tests/tree/test_directory_tree_reload_node.py:18]: I did both break it down into two cases and added some comments. Let me know if you still think the tests should be improved.

File: src/textual/widgets/_directory_tree.py [hunk 6]
```diff
@@ -146,6 +149,56 @@ def reload(self) -> None:
         # loading up.
         self._add_to_load_queue(self.root)
 
+    def clear_subtree(self, node: TreeNode[DirEntry]) -> Self:
+        """Clear all nodes under the given node.
+
+        Returns:
+            The `Tree` instance.
+        """
+        self._line_cache.clear()
```
[the remark this answers is not in the document]
Comment (gergely-elias) [reply src/textual/widgets/_directory_tree.py]: I noticed the base class already has an `invalidate_cache` method, which needs just a little bit more than needed here, so in the end I did not use that one. On the other hand - as far as I can see - calling that one would also be an option here, so maybe you'd prefer to be this done with that.

Also I did not implement a utility method to increase `self._updates`, looked like somewhat of an overkill. But again, let me know if you'd prefer it being done like that.

**Review by rodrigogiraoserrao — APPROVED** [review]
After addressing Will's comments, the PR looks good to me.

Sorry for the time you've been waiting. We'll come back to you ASAP.

**Comment by gergely-elias** [comment]
@rodrigogiraoserrao no worries
I think functionality wise this PR is complete, but am not sure if I really follow all your design conventions - so if reorganizing the parts regarding cache clearing is desired, please don't hesitate to tell

File: src/textual/widgets/_tree.py [hunk 7]
```diff
@@ -692,14 +692,18 @@ def get_label_width(self, node: TreeNode[TreeDataType]) -> int:
         label = self.render_label(node, NULL_STYLE, NULL_STYLE)
         return label.cell_len
 
+    def clear_line_cache(self) -> None:
```
Comment (willmcgugan) [inline src/textual/widgets/_tree.py:695]: This should be private, i.e. prefixed with `_`. It's not something that the dev needs to know about generally.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Looks good. Just one super trivial request.

**Comment by gergely-elias** [comment]
@rodrigogiraoserrao @willmcgugan
all review requests have been addressed, please approve

**Comment by rodrigogiraoserrao** [comment]
@gergely-elias thank you so much for all your work and patience!
I'm waiting for the tests to pass ([relevant xkcd](https://xkcd.com/303/)) and I'll merge it :)

**Review by willmcgugan — APPROVED** [review]
LGTM!

**Pull request merged**